scrypto-unit/src/test_runner.rs

```diff
@@ -2304,8 +2312,18 @@ impl StateHashSupport {
         }
     }
 
-    pub fn update_with(&mut self, db_updates: &DatabaseUpdates) {
+    pub fn update_with(
+        &mut self,
+        db_updates: &DatabaseUpdates,
+        partition_deletions: &IndexSet<DbPartitionKey>,
+    ) {
         let mut hash_changes = Vec::new();
+        // TODO(during review): this assumes that partition deletions happen before other changes - rightfully?
```

Comment:
Yes, we can have this as an invariant. Perhaps to emphasize this fact we should remove the `HashChange` enum and just accept two parameters in `put_at_next_version`, one for batch changes (to be processed first) and another for substate hash changes.

Reply:
huh, I just re-read this: 
https://github.com/radixdlt/radixdlt-scrypto/blob/0fe0b31b720a501d2785e50aa6cdc5276f2f3eff/radix-engine/src/track/track.rs#L24-L26

It currently says `after`, but it feels more natural to apply batch changes first (because only then the "small" changes have opportunity to adjust the state after the batch change). In general, I think it's worth addressing the `TODO: ... merging it into database/system updates` ~now too. I guess we'll discuss it today.

Reply:
> Perhaps to emphasize this fact we should remove the `HashChange` enum and just accept two parameters in `put_at_next_version`, one for batch changes (to be processed first) and another for substate hash changes.

This was my prototype-v1 impl here. (a smaller change than `HashChange` enum)

However, it turned out to be misaligned with JMT's API. We need to call the underlying JMT `batch_put_value_set()` exactly once per state version. So, I still would _not_ be able to do anything like:
```
update_jmt(batch_changes);
update_jmt(substate_changes);
```
_(I'd still have to merge these 2 lists into a single struct "indexed" by Node:Partition:Substate)_

So, since the "2 fields" API wouldn't bring me any convenience, I figured it's better to go with a "natural" API, i.e. literally a list of changes to be applied to the storage, in the given order.